### PR TITLE
Refs #21160 -- Replaced DatabaseFeatures.supports_1000_query_parameters by a DatabaseFeatures.max_query_params.

### PR DESCRIPTION
**Pull request by felixxm**
Ticket [21160](https://code.djangoproject.com/ticket/21160).

### CODE REVIEW

File: tests/queries/tests.py [hunk 1]
```diff
@@ -2523,32 +2523,34 @@ def test_null_ordering_added(self):
...
 
-    # Sqlite 3 does not support passing in more than 1000 parameters except by
-    # changing a parameter at compilation time.
-    @skipUnlessDBFeature('supports_1000_query_parameters')
     def test_ticket14244(self):
         # The "in" lookup works with lists of 1000 items or more.
         # The numbers amount is picked to force three different IN batches
         # for Oracle, yet to be less than 2100 parameter limit for MSSQL.
         numbers = list(range(2050))
         Number.objects.all().delete()
```
Comment (charettes) [inline tests/queries/tests.py]: Not sure why this line is required, we don't seem to be creating any `Number` in `setupTestData()` 🤷‍♂️

File: tests/queries/tests.py [hunk 2]
```diff
@@ -2523,32 +2523,34 @@ def test_null_ordering_added(self):
         self.assertEqual(sql.find(fragment, pos + 1), -1)
         self.assertEqual(sql.find("NULL", pos + len(fragment)), pos + len(fragment))
 
-    # Sqlite 3 does not support passing in more than 1000 parameters except by
-    # changing a parameter at compilation time.
-    @skipUnlessDBFeature('supports_1000_query_parameters')
     def test_ticket14244(self):
```
Comment (charettes) [inline tests/queries/tests.py]: Might be worth giving this test a more significant name?

File: tests/queries/tests.py (hunk 1 above)
Comment (felixxm) [inline tests/queries/tests.py]: You're right. It's very old line 👴.

**Comment by felixxm** [comment]
@charettes Thanks for review.

**Comment by felixxm** [comment]
From the official [3rd-party-database-backends](https://docs.djangoproject.com/en/1.10/ref/databases/#using-a-3rd-party-database-backend) only [ODBC](https://github.com/lionheart/django-pyodbc) is using the  `supports_1000_query_parameters` feature:
```
tests/django16/queries/tests.py
tests/django14/queries/tests.py
tests/django15/queries/tests.py
```
but only in tests based on the old Django versions.

File: django/db/backends/base/features.py [hunk 3]
```diff
@@ -90,8 +90,8 @@ class BaseDatabaseFeatures:
     # Does the backend order NULL values as largest or smallest?
     nulls_order_largest = False
 
-    # Is there a 1000 item limit on query parameters?
-    supports_1000_query_parameters = True
+    # Is there a query parameters limit?
```
Comment (timgraham) [inline django/db/backends/base/features.py]: `# The database's limit on the number of query parameters.`

File: tests/queries/tests.py [hunk 4]
```diff
@@ -2523,32 +2523,33 @@ def test_null_ordering_added(self):
...
-            Number.objects.filter(num__in=numbers[:2000]).count(),
-            2000
-        )
-        self.assertEqual(
-            Number.objects.filter(num__in=numbers).count(),
-            len(numbers)
-        )
+        max_query_params = connection.features.max_query_params or len(numbers)
+        if max_query_params >= len(numbers):
+            self.assertEqual(
```
Comment (timgraham) [inline tests/queries/tests.py]: These assertions can be single lined.

File: tests/queries/tests.py [hunk 5]
```diff
@@ -2523,32 +2523,33 @@ def test_null_ordering_added(self):
...
-        self.assertEqual(
-            Number.objects.filter(num__in=numbers[:2000]).count(),
-            2000
-        )
-        self.assertEqual(
-            Number.objects.filter(num__in=numbers).count(),
-            len(numbers)
-        )
+        max_query_params = connection.features.max_query_params or len(numbers)
+        if max_query_params >= len(numbers):
```
Comment (timgraham) [inline tests/queries/tests.py]: Looking at the original patch (320c46999c), I'm not sure if the if/elif structure is correct. Assuming `max_query_params` is used only on SQLite are all branches used?

Comment (felixxm) [inline tests/queries/tests.py]: You're right. Due to the fact that `max_query_params` is used only on SQLite I think that we can simplify this test.

**Comment by felixxm** [comment]
@timgraham Thanks for review.

File: tests/queries/tests.py [hunk 6]
```diff
@@ -2523,32 +2523,16 @@ def test_null_ordering_added(self):
...
```
Comment (timgraham) [inline tests/queries/tests.py]: Would this be clearer?
```
max_query_params = connection.features.max_query_params
if max_query_params is None or max_query_params >= len(numbers):
```

File: tests/queries/tests.py [hunk 7]
```diff
@@ -2523,32 +2523,16 @@ def test_null_ordering_added(self):
...
+            Number.objects.bulk_create(Number(num=num) for num in numbers)
+            for number in [1000, 1001, 2000, len(numbers)]:
+                self.assertEqual(Number.objects.filter(num__in=numbers[:number]).count(), number)
```
Comment (timgraham) [inline tests/queries/tests.py]: You could use `subtest()` for the loop.

**Pull request merged**